File: submissions/0838-design-linked-list/solution.c
typedef struct Node {
    int val;
    struct Node* prev;
    struct Node* next;
} Node;

typedef struct {
    Node* preHead;
    size_t size;
} MyLinkedList;

static Node* mkNode(int val, Node* restrict prev, Node* restrict next) {
    Node* node = calloc(1, sizeof *node);
    if (!node) return NULL;

    node->val = val;
    node->prev = prev;
    node->next = next;
    return node;
}

MyLinkedList* myLinkedListCreate(void) {
    MyLinkedList* list = calloc(1, sizeof *list);
    if (!list) return NULL;

    Node* preHead = mkNode(0, NULL, NULL);
    if (!preHead) {
        free(list);
        return NULL;
    }

    Node* head = mkNode(0, NULL, NULL);
    if (!head) {
        free(list);
        free(preHead);
        return NULL;
    }

    preHead->next = head;
    head->prev = preHead;
    list->preHead = preHead;

    return list;
}

int myLinkedListGet(MyLinkedList* obj, size_t index) {
    if (!obj || index >= obj->size) return -1;

    Node* cur = obj->preHead->next;
    for (size_t i = 0; i < index; ++i) {
        cur = cur->next;
    }
    return cur->val;
}

void myLinkedListAddAtIndex(MyLinkedList* obj, size_t index, int val) {
    if (!obj || index > obj->size) return;

    Node* prev = obj->preHead;
    for (size_t i = 0; i < index; ++i) {
        prev = prev->next;
    }
    Node* next = prev->next;
    Node* newNode = mkNode(val, prev, next);
    if (!newNode) return;

    prev->next = newNode;
    next->prev = newNode;

    obj->size++;
}

void myLinkedListAddAtHead(MyLinkedList* obj, int val) {
    myLinkedListAddAtIndex(obj, 0, val);
}

void myLinkedListAddAtTail(MyLinkedList* obj, int val) {
    if (!obj) return;
    myLinkedListAddAtIndex(obj, obj->size, val);
}

void myLinkedListDeleteAtIndex(MyLinkedList* obj, size_t index) {
    if (!obj || index >= obj->size) return;

    Node* prev = obj->preHead;
    for (size_t i = 0; i < index; ++i) {
        prev = prev->next;
    }
    Node* toDelete = prev->next;
    if (!toDelete) return;
    Node* next = toDelete->next;

    prev->next = next;
    next->prev = prev;

    free(toDelete);
    obj->size--;
}

void myLinkedListFree(MyLinkedList* obj) {
    if (!obj) return;

    Node* cur = obj->preHead;
    while (cur) {
        Node* next = cur->next;
        free(cur);
        cur = next;
    }
    free(obj);
}

File: submissions/0838-design-linked-list/solution.java
class MyLinkedList {
    private Node head;
    private Node tail;
    private int length;

    public MyLinkedList() {
        length = 0;
    }

    public int get(int index) {
        final var node = _get(index);
        return node == null ? -1 : node.val;
    }

    public void addAtHead(int val) {
        final var newHead = new Node(val, null, head);
        if (head != null) {
            head.prev = newHead;
        }
        head = newHead;
        if (length == 0) tail = head;
        length++;
    }

    public void addAtTail(int val) {
        if (length == 0) {
            addAtHead(val);
            return;
        }
        final var newTail = new Node(val, tail, null);
        tail.next = newTail;
        tail = newTail;
        length++;
    }

    public void addAtIndex(int index, int val) {
        if (index < 0) return;
        if (index > length) return;
        if (index == 0) {
            addAtHead(val);
            return;
        }
        if (index == length) {
            addAtTail(val);
            return;
        }

        final Node cur = _get(index - 1);
        if (cur == null) return;
        final var node = new Node(val, cur, cur.next);
        cur.next = node;
        node.next.prev = node;
        length++;
    }

    public void deleteAtIndex(int index) {
        final var node = _get(index);
        if (node == null) return;

        final var prev = node.prev;
        final var next = node.next;
        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
        }
        if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
        }

        length--;
    }

    private Node _get(int index) {
        if (index < 0 || index >= length) return null;

        Node cur = head;
        for (int i = 0; i < index; i++) {
            cur = cur.next;
        }
        return cur;
    }
}

class Node {
    int val;
    Node prev;
    Node next;

    public Node(int val, Node prev, Node next) {
        this.val = val;
        this.prev = prev;
        this.next = next;
    }
}

